### PR TITLE
make smoke tests Java 8 compatible

### PR DESCRIPTION
**Pull request by sbera87**
*Issue #, if available:*

*Description of changes:
Previous iteration of changes are based on Java 16 which is not compatible with cdk. Hence, this change is necessary.

*Check all that applies:*
- [x] Did a review by yourself.
- [ ] Added proper tests to cover this PR. (If tests are not applicable, explain.)
- [ ] Checked if this PR is a breaking (APIs have been changed) change.
- [ ] Checked if this PR will _not_ introduce cross-platform inconsistent behavior.
- [ ] Checked if this PR would require a ReadMe/Wiki update.

Check which platforms you have built SDK on to verify the correctness of this PR.
- [x] Linux
- [ ] Windows
- [ ] Android
- [x] MacOS
- [ ] IOS
- [ ] Other Platforms

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsCMakeWriter.java [hunk 1]
```diff
@@ -19,58 +19,55 @@ public String generate()
...
+                folderNamespace + "-smoke-tests\n" +
+                "\"Tests for the AWS " + projectNameCaps + " C++ SDK\"\n" +
+                "testing-resources\n" +
+                "aws-cpp-sdk-" + folderNamespace + "\n" +
+                "aws-cpp-sdk-core\n" +
+                ")"
         );
 
-        write(""" 
-            file(GLOB AWS_$L_GENERATED_SMOKE_TEST_SRC
```
Comment (sbera87) [inline tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmokeTestsCMakeWriter.java:33]: Multi line strings not supported on Java 8. Hence needed refactoring

File: tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts [hunk 2]
```diff
@@ -56,44 +58,49 @@ tasks.register("generate-smithy-build") {
...
                     // as part of the actual build task done by the smithy gradle plugin.
                     .assemble().result.get();
-            val services = model.shapes(ServiceShape::class.javaObjectType).sorted().toList();
+
+            val services = model.shapes(ServiceShape::class.java).sorted().toList();
+
             if (services.size != 1) {
-                throw Exception("There must be exactly one service in each aws model file, but found " +
-                        "${services.size} in ${file.name}: ${services.map { it.id }}");
+                println("Unexpected number of service shapes detected: ${services.size} while processing file: ${file.name}. ")
```
Comment (sbera87) [inline tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts:72]: deliberately not throwing exception to allow other services to move on

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java [hunk 3]
```diff
@@ -17,9 +17,16 @@ public class SmithyC2JNamespaceMap {
 
     private SmithyC2JNamespaceMap(String jsonString){
 
-        Type mapType = new TypeToken<Map<String, String>>(){}.getType();
-        Gson gson = new Gson();
-        this.SMITHY_C2J_SERVICE_NAME_MAP = gson.fromJson(jsonString, mapType);
+        if(jsonString.isEmpty())
```
Comment (sbera87) [inline tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java:20]: This is a fix when no mapping is provided

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppImportContainer.java [hunk 4]
```diff
@@ -41,21 +43,19 @@ public CppImportContainer(String namespace) {
             "algorithm"
             );
 
-        this.unitTestHeaders = Set.of(
-            "aws/testing/AwsCppSdkGTestSuite.h",
-            "aws/testing/AwsTestHelpers.h"
-            );
+        this.unitTestHeaders = new HashSet<>();
```
Comment (sbiscigl) [inline tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppImportContainer.java:46]: Should use guavas `ImmubtableSet.of` and add the [guava dep](https://mvnrepository.com/artifact/com.google.guava/guava) if needed, would be nice but not required

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppImportContainer.java [hunk 5]
```diff
@@ -41,21 +43,19 @@ public CppImportContainer(String namespace) {
...
-            Map.entry("Aws::String", "aws/core/utils/memory/stl/AWSString.h"),
-            Map.entry("Aws::Map", "aws/core/utils/memory/stl/AWSMap.h"),
-            Map.entry("Aws::Utils::DateTime", "aws/core/utils/DateTime.h"),
-            Map.entry("Aws::Utils::Document","aws/core/utils/Document.h"),
-            Map.entry("Aws::Utils::ByteBuffer","aws/core/utils/Array.h")
-        );
-        
-        dynamicHeaders.add(String.format("aws/%s/%sClient.h", c2jNamespace, clientNamespace));
+         //this will be added to based upon datastructures found
+        containerHeaderMap = new HashMap<>();
```
Comment (sbiscigl) [inline tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppImportContainer.java:51]: Similar as above should use guavas 'ImmutableMap.of' and add the dep

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppSymbolVisitor.java [hunk 6]
```diff
@@ -5,6 +5,7 @@
 
  package com.amazonaws.util.awsclientsmithygenerator.generators;
 
+ import lombok.val;
```
Comment (sbiscigl) [inline tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppSymbolVisitor.java:8]: Dont use Lombok val, spell out the complete type

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java (hunk 3 above)
Comment (sbiscigl) [reply tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java:20]: when would there be no mapping file, dont we always pass it one? shouldn't this be an exception?

File: tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts (hunk 2 above)
Comment (sbiscigl) [reply tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts:72]: we should catch and throw issues when we find them, and leave them up to the caller, i.e. the CI script to ignore. for instance here, why are there more than 1 service shape? what is wrong with the smithy files that allows this to be the case?

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java (hunk 3 above)
Comment (sbera87) [reply tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java:20]: This should work with or without mapping. Mapping is just an override. Codegen script passes mapping but it doesnt stop anyone from just using gradle build without one for testing (where no service needs to be remapped) will cause the issue

File: tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts (hunk 2 above)
Comment (sbera87) [reply tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts:72]: yes, but it affects all the other services that don't have issue. I can still throw it but it is going to show incomplete builds in pipeline if even one service has an issue

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java (hunk 3 above)
Comment (sbiscigl) [reply tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/SmithyC2JNamespaceMap.java:20]: yeah thats fair

File: tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts (hunk 2 above)
Comment (sbiscigl) [reply tools/code-generation/smithy/codegen/cpp-smoke-tests/build.gradle.kts:72]: the doesn't have anything to do with the pipeline, or CI at all, at pure level, if the code generator fails, it should bubble up the error to the caller, and the caller is responsible for handling it. in this case the CI pipeline.

> for instance here, why are there more than 1 service shape? what is wrong with the smithy files that allows this to be the case?

you didnt answer this question -- why do we need this and is this a data problem. what services cause this problem and why.

File: tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppSymbolVisitor.java (hunk 6 above)
Comment (sbera87) [reply tools/code-generation/smithy/codegen/cpp-smoke-tests-codegen/src/main/java/com/amazonaws/util/awsclientsmithygenerator/generators/CppSymbolVisitor.java:8]: ok, that was IDE suggestion. will change

**Review by SergeyRyabinin — APPROVED** [review]

**Pull request merged**